### PR TITLE
Fix thread-unsafety with custom operations

### PR DESCRIPTION
**Pull request by gadmm**
This PR fixes thread-unsafety of custom operations table. It used to store custom operations in global non-atomic linked lists. This PR pushes onto the list atomically. A CAS is enough because nothing is ever removed from the lists.

I think the present straightforward approach is fine, but I am open to an alternative based on lock-free skiplists which is a tad more work. (You need to initialize the skip-lists somewhere; and probably do not need to clean-up garbage regularly.)

A better-documented version is available at https://github.com/gadmm/ocaml/compare/custom_thread_safety2..custom_thread_safety but this style needs the listed (minor) change to `platform.h`.

### CODE REVIEW

**Review by kayceesrk — APPROVED** [review]
LGTM. The PR fixes the thread unsafety correctly.

**Review by gasche — APPROVED** [review]
I also agree that this is correct and it even makes the code slightly nicer / more factorized. Thanks!

Note: I cannot figure out what the changes in platform.h in [your other version](https://github.com/gadmm/ocaml/compare/custom_thread_safety2..custom_thread_safety) are doing and why they are needed. My eyes just tell me that you reimplemented inline functions as macros of the exact same name. (Note note: looking at this reminded me that I wish `store_rel` was called `stored_release` to be less easy to confuse with `store_relaxed` for non-experts.)

**Comment by gadmm** [comment]
It did not type-check with the previous version of `platform.h`. Let me know if I should push the alternative version (better-documented to my taste) including the change to `platform.h`. (And I can throw in a search&replace of store_rel into store_release if you want)

**Comment by kayceesrk** [comment]
> Let me know if I should push the alternative version (better-documented to my taste) including the change to platform.h. (And I can throw in a search&replace of store_rel into store_release if you want)

This is best done as a separate cleanup PR.

**Comment by gadmm** [comment]
Updated `Changes` and squashed.

**Comment by gadmm** [comment]
@Octachron This is candidate for backporting, if there is a 5.0.1 planned.

**Comment by gadmm** [comment]
> This is best done as a separate cleanup PR.

I agree but I also have not had a good experience with cleanup PRs among others. So in case a core developer decides to make the change rel->release and acq->acquire, please also consider to include the change to platform.h which is a welcome usability improvement!

**Comment by gasche** [comment]
> I also have not had a good experience with cleanup PRs among others.

I think there is a difference between "monster bugfix&cleanup PRs that takes weeks and aspirine to review" and "small cleanup PR that fixes a clearly identified issue". But do as you prefer. (I don't have the energy to adopt your change suggestion.)

> @Octachron This is candidate for backporting, if there is a 5.0.1 planned.

Right now the plan is rather to release 5.1 quickly. (In this context of not knowing whether a 5.0.1 will occur, I would focus on considering backporting for hot bugfixes that clearly affect several users.)

**Comment by gadmm** [comment]
Under cygwin. List of failed tests:
    tests/lib-dynlink-initializers/'test1_main.ml' with 1.2.1.4 (ocamlopt.byte)

I'll let you merge if this is a known false-positive.

**Comment by kayceesrk** [comment]
Completely agree with @gasche. Clearly motivated, well-contained cleanup PRs are easy to review and approve. Mixing bug fixes and cleanups in the same PRs makes it harder to review. As an example,

>  rel->release and acq->acquire, please also consider to include the change to platform.h which is a welcome usability improvement!

Would be an improvement and an easy PR to review.

**Comment by gasche** [comment]
There is a weird CI failure on Windows, and the output is a mouthful:

```
Commandline: C:\projects\🐫реализация-mingw64\runtime\ocamlrun.exe C:\projects\🐫реализация-mingw64\ocamlopt.exe -I C:\projects\🐫реализация-mingw64\runtime -nostdlib -I C:\projects\🐫реализация-mingw64\stdlib -I C:\projects\🐫реализация-mingw64\otherlibs\dynlink -I C:\projects\🐫реализация-mingw64\otherlibs\dynlink\native dynlink.cmxa -o C:/projects/🐫реализация-mingw64/testsuite/tests/lib-dynlink-initializers/_ocamltest\tests/lib-dynlink-initializers\test1_main\ocamlopt.byte/test1.exe test1_main.cmx test1_inited_second.cmx
[...]
** Fatal error: Cannot run cygpath -m "libshell32" "libshell32.lib" "libshell32.dll.a" "libshell32.a" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\native\libshell32" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\native\libshell32.lib" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\native\libshell32.dll.a" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\native\libshell32.a" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\libshell32" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\libshell32.lib" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\libshell32.dll.a" "C:\projects\🐫реализация-mingw64\otherlibs\dynlink\libshell32.a" "C:\projects\🐫реализация-mingw64\stdlib\libshell32" "C:\projects\🐫реализация-mingw64\stdlib\libshell32.lib" "C:\projects\🐫реализация-mingw64\stdlib\libshell32.dll.a" "C:\projects\🐫реализация-mingw64\stdlib\libshell32.a" "C:\projects\🐫реализация-mingw64\runtime\libshell32" "C:\projects\🐫реализация-mingw64\runtime\libshell32.lib" "C:\projects\🐫реализация-mingw64\runtime\libshell32.dll.a" "C:\projects\🐫реализация-mingw64\runtime\libshell32.a" "C:/Program Files/Бактріан🐫/lib/ocaml\flexdll\libshell32" "C:/Program Files/Бактріан🐫/lib/ocaml\flexdll\libshell32.lib" "C:/Program Files/Бактріан🐫/lib/ocaml\flexdll\libshell32.dll.a" "C:/Program Files/Бактріан🐫/lib/ocaml\flexdll\libshell32.a" "C:\projects\🐫реализация-mingw64\testsuite\..\stdlib\flexdll\libshell32" "C:\projects\🐫реализация-mingw64\testsuite\..\stdlib\flexdll\libshell32.lib" "C:\projects\🐫реализация-mingw64\testsuite\..\stdlib\flexdll\libshell32.dll.a" "C:\projects\🐫реализация-mingw64\testsuite\..\stdlib\flexdll\libshell32.a" "/usr/lib/gcc/x86_64-w64-mingw32/11/libshell32" "/usr/lib/gcc/x86_64-w64-mingw32/11/libshell32.lib" "/usr/lib/gcc/x86_64-w64-mingw32/11/libshell32.dll.a" "/usr/lib/gcc/x86_64-w64-mingw32/11/libshell32.a" "/usr/x86_64-w64-mingw32/lib/x86_64-w64-mingw32/11/libshell32" "/usr/x86_64-w64-mingw32/lib/x86_64-w64-mingw32/11/libshell32.lib" "/usr/x86_64-w64-mingw32/lib/x86_64-w64-mingw32/11/libshell32.dll.a" "/usr/x86_64-w64-mingw32/lib/x86_64-w64-mingw32/11/libshell32.a" "/usr/x86_64-w64-mingw32/lib/libshell32" "/usr/x86_64-w64-mingw32/lib/libshell32.lib" "/usr/x86_64-w64-mingw32/lib/libshell32.dll.a" "/usr/x86_64-w64-mingw32/lib/libshell32.a" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/x86_64-w64-mingw32/11/libshell32" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/x86_64-w64-mingw32/11/libshell32.lib" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/x86_64-w64-mingw32/11/libshell32.dll.a" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/x86_64-w64-mingw32/11/libshell32.a" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/libshell32" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/libshell32.lib" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/libshell32.dll.a" "/usr/x86_64-w64-mingw32/sys-root/mingw/lib/libshell32.a"
File "caml_startup", line 1:
Error: Error during linking (exit code 2)
```

@dra27, would you per chance magically know if this has any chance to be related to the current issue or can be set aside for now?

**Comment by dra27** [comment]
It can be set aside (although I just restarted AppVeyor) - it's the failure of a command in Cygwin to translate a path. Either the issue will magically resolve itself in a Cygwin update (not impossible) or I'll do battle with flexlink at some point and either harden the call to it or eliminate the need to do it. The problem is that it it's _very_ sporadic, so it's hard to be certain one's actually fixed it.

**Comment by gasche** [comment]
Thanks! We wish you the very best cygwin paths in 2023.

**Pull request merged**
